core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,780 @@
+        transitPartitionChunkState(aug, State.STANDBY, State.VISIBLE);
+      }
+    }
+  }
+
+  private void addTo(AtomicUpdateGroup<T> aug, State state)
+  {
+    final AtomicUpdateGroup<T> existing = getStateMap(state)
+        .computeIfAbsent(RootPartitionRange.of(aug), k -> createMinorVersionToAugMap(state))
+        .put(aug.getMinorVersion(), aug);
+
+    if (existing != null) {
+      throw new ISE("AtomicUpdateGroup[%s] is already in state[%s]", aug, state);
+    }
+
+    handleAdd(aug, state);
+  }
+
+  public void add(PartitionChunk<T> chunk)
+  {
+    final PartitionChunk<T> existingChunk = knownPartitionChunks.put(chunk.getChunkNumber(), chunk);
+    if (existingChunk != null && !existingChunk.equals(chunk)) {
+      throw new ISE(
+          "existingChunk[%s] is different from newChunk[%s] for partitionId[%d]",
+          existingChunk,
+          chunk,
+          chunk.getChunkNumber()
+      );
+    }
+
+    // Find atomicUpdateGroup of the new chunk
+    AtomicUpdateGroup<T> atomicUpdateGroup = searchForStateOf(chunk, State.OVERSHADOWED);
+
+    if (atomicUpdateGroup != null) {
+      atomicUpdateGroup.add(chunk);
+    } else {
+      atomicUpdateGroup = searchForStateOf(chunk, State.STANDBY);
+
+      if (atomicUpdateGroup != null) {
+        atomicUpdateGroup.add(chunk);
+        handleAdd(atomicUpdateGroup, State.STANDBY);
+      } else {
+        atomicUpdateGroup = searchForStateOf(chunk, State.VISIBLE);
+
+        if (atomicUpdateGroup != null) {
+          // A new chunk of the same major version and partitionId can be added in segment handoff
+          // from stream ingestion tasks to historicals
+          final PartitionChunk<T> existing = atomicUpdateGroup.replaceChunkWith(chunk);
+          if (existing == null) {
+            throw new ISE(
+                "Can't add a new partitionChunk[%s] to a visible atomicUpdateGroup[%s]",
+                chunk,
+                atomicUpdateGroup
+            );
+          } else if (!chunk.equals(existing)) {
+            throw new ISE(
+                "WTH? a new partitionChunk[%s] has the same partitionId but different from existing chunk[%s]",
+                chunk,
+                existing
+            );
+          }
+        } else {
+          final AtomicUpdateGroup<T> newAtomicUpdateGroup = new AtomicUpdateGroup<>(chunk);
+
+          // Decide the initial state of the new atomicUpdateGroup
+          final boolean overshadowed = visibleGroup
+              .values()
+              .stream()
+              .flatMap(map -> map.values().stream())
+              .anyMatch(group -> group.isOvershadow(newAtomicUpdateGroup));
+
+          if (overshadowed) {
+            addTo(newAtomicUpdateGroup, State.OVERSHADOWED);
+          } else {
+            addTo(newAtomicUpdateGroup, State.STANDBY);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Handles of removal of an empty atomicUpdateGroup from a state.
+   */
+  private void handleRemove(
```

Comment:
Similarly for the remove methods, suggest renaming:

`remove` -> `removeChunk`
`tryRemoveFromState` -> `tryRemoveChunkFromGroupWithState`
`handleRemove` -> `determineVisibleGroupAfterRemove`

Reply:
Renamed.